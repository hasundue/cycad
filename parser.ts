/**
 * A sub-module of cycad_sitter which provides a parser without async operations.
 *
 * ## Example
 *
 * ```ts
 * import { parser } from "https://deno.land/x/cycad_sitter{version}/parser.ts?lang=markdown";
 *
 * const tree = parser.parse(`Hello, World!`);
 * console.log(tree.rootNode.toString());
 * ```
 *
 * @module
 */

import { isLanguage } from "./src/langs.generated.ts";
import { Parser } from "./src/parser.ts";

export { Parser } from "./src/parser.ts";

const lang = new URL(import.meta.url).searchParams.get("lang");

if (!lang) {
  throw new Error(`Missing language parameter in URL: ${import.meta.url}`);
}

if (!isLanguage(lang)) {
  throw new TypeError(`Unknown language: ${lang}`);
}

export const parser = await Parser.create(lang);
