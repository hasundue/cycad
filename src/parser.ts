import { exists } from "https://deno.land/std@0.204.0/fs/exists.ts";
import TreeSitter from "npm:web-tree-sitter@0.20.8";
import { isLanguage } from "./langs.generated.ts";
import { getLanguagePath } from "./paths.ts";
import { buildParser } from "./build.ts";

export interface Parser extends TreeSitter {
  parse(
    input: string | TreeSitter.Input,
    previousTree?: TreeSitter.Tree,
    options?: TreeSitter.Options,
  ): TreeSitter.Tree;
  reset(): void;
  getLogger(): TreeSitter.Logger;
  setLogger(logFunc?: TreeSitter.Logger | undefined | null): void;
  setTimeoutMicros(value: number): void;
  getTimeoutMicros(): number;
}

await TreeSitter.init();

const lang = new URL(import.meta.url).hash.slice(1);
console.log(lang);

if (!isLanguage(lang)) {
  throw new TypeError(`Unknown language: ${lang}`);
}

const wasm = getLanguagePath(lang);

if (!(await exists(wasm))) {
  await buildParser(lang);
}

const Lang = await TreeSitter.Language.load(wasm);

const _parser = new TreeSitter();
_parser.setLanguage(Lang);

export const parser: Parser = _parser;
