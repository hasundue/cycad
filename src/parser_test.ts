import {
  assertSnapshot,
} from "https://deno.land/std@0.204.0/testing/snapshot.ts";
import { parser } from "./parser.ts#markdown";

Deno.test("parser", async (t) => {
  const src = await Deno.readTextFile("README.md");
  const tree = parser.parse(src);
  await assertSnapshot(t, tree.rootNode.toString());
});
