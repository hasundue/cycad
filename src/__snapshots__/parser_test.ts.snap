export const snapshot = {};

snapshot[`parser 1`] = `"(document (section (atx_heading (atx_h1_marker) heading_content: (inline)) (paragraph (inline)) (section (atx_heading (atx_h2_marker) heading_content: (inline)) (list (list_item (list_marker_minus) (paragraph (inline))) (list_item (list_marker_minus) (paragraph (inline (block_continuation)))) (list_item (list_marker_minus) (paragraph (inline (block_continuation))))))))"`;
